<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Recover intra-sample doublets — recoverDoublets • scDblFinder</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Recover intra-sample doublets — recoverDoublets"><meta name="description" content="Recover intra-sample doublets that are neighbors to known inter-sample doublets in a multiplexed experiment."><meta property="og:description" content="Recover intra-sample doublets that are neighbors to known inter-sample doublets in a multiplexed experiment."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDblFinder</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.19.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/scDblFinder.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/computeDoubletDensity.html">Scoring potential doublets from simulated densities</a></li>
    <li><a class="dropdown-item" href="../articles/findDoubletClusters.html">Detecting clusters of doublet cells with DE analyses</a></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Introduction to the scDblFinder package</a></li>
    <li><a class="dropdown-item" href="../articles/recoverDoublets.html">Recovering intra-sample doublets</a></li>
    <li><a class="dropdown-item" href="../articles/scATAC.html">Doublet identifiation in single-cell ATAC-seq</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/plger/scDblFinder/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Recover intra-sample doublets</h1>
      <small class="dont-index">Source: <a href="https://github.com/plger/scDblFinder/blob/devel/R/recoverDoublets.R" class="external-link"><code>R/recoverDoublets.R</code></a></small>
      <div class="d-none name"><code>recoverDoublets.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Recover intra-sample doublets that are neighbors to known inter-sample doublets in a multiplexed experiment.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">recoverDoublets</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'ANY'</span></span>
<span><span class="fu">recoverDoublets</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">doublets</span>,</span>
<span>  <span class="va">samples</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  transposed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  subset.row <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  BNPARAM <span class="op">=</span> <span class="fu">KmknnParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">SerialParam</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SummarizedExperiment'</span></span>
<span><span class="fu">recoverDoublets</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span>, assay.type <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SingleCellExperiment'</span></span>
<span><span class="fu">recoverDoublets</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span>, use.dimred <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A log-expression matrix for all cells (including doublets) in columns and genes in rows.
If <code>transposed=TRUE</code>, this should be a matrix of low-dimensional coordinates where each row corresponds to a cell.</p>
<p>Alternatively, a <a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a> or <a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a> containing
(i) a log-expression matrix in the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></code> as specified by <code>assay.type</code>,
or (ii) a matrix of reduced dimensions in the <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDims</a></code> as specified by <code>use.dimred</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>For the generic, additional arguments to pass to specific methods.</p>
<p>For the SummarizedExperiment method, additional arguments to pass to the ANY method.</p>
<p>For the SingleCellExperiment method, additional arguments to pass to the SummarizedExperiment method.</p></dd>


<dt id="arg-doublets">doublets<a class="anchor" aria-label="anchor" href="#arg-doublets"></a></dt>
<dd><p>A logical, integer or character vector specifying which cells in <code>x</code> are known (inter-sample) doublets.</p></dd>


<dt id="arg-samples">samples<a class="anchor" aria-label="anchor" href="#arg-samples"></a></dt>
<dd><p>A numeric vector containing the relative proportions of cells from each sample,
used to determine how many cells are to be considered as intra-sample doublets.</p></dd>


<dt id="arg-k">k<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>Integer scalar specifying the number of nearest neighbors to use for computing the local doublet proportions.</p></dd>


<dt id="arg-transposed">transposed<a class="anchor" aria-label="anchor" href="#arg-transposed"></a></dt>
<dd><p>Logical scalar indicating whether <code>x</code> is transposed, i.e., cells in the rows.</p></dd>


<dt id="arg-subset-row">subset.row<a class="anchor" aria-label="anchor" href="#arg-subset-row"></a></dt>
<dd><p>A logical, integer or character vector specifying the genes to use for the neighbor search.
Only used when <code>transposed=FALSE</code>.</p></dd>


<dt id="arg-bnparam">BNPARAM<a class="anchor" aria-label="anchor" href="#arg-bnparam"></a></dt>
<dd><p>A BiocNeighborParam object specifying the algorithm to use for the nearest neighbor search.</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>A <a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html" class="external-link">BiocParallelParam</a> object specifying the parallelization to use for the nearest neighbor search.</p></dd>


<dt id="arg-assay-type">assay.type<a class="anchor" aria-label="anchor" href="#arg-assay-type"></a></dt>
<dd><p>A string specifying which assay values contain the log-expression matrix.</p></dd>


<dt id="arg-use-dimred">use.dimred<a class="anchor" aria-label="anchor" href="#arg-use-dimred"></a></dt>
<dd><p>A string specifying whether existing values in <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDims</a>(x)</code> should be used.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A DataFrame containing one row per cell and the following fields:</p><ul><li><p><code>proportion</code>, a numeric field containing the proportion of neighbors that are doublets.</p></li>
<li><p><code>known</code>, a logical field indicating whether this cell is a known inter-sample doublet.</p></li>
<li><p><code>predicted</code>, a logical field indicating whether this cell is a predicted intra-sample doublet.</p></li>
</ul><p>The <code>metadata</code> contains <code>intra</code>, a numeric scalar containing the expected number of intra-sample doublets.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>In multiplexed single-cell experiments, we can detect doublets as libraries with labels for multiple samples.
However, this approach fails to identify doublets consisting of two cells with the same label.
Such cells may be problematic if they are still sufficiently abundant to drive formation of spurious clusters.</p>
<p>This function identifies intra-sample doublets based on the similarity in expression profiles to known inter-sample doublets.
For each cell, we compute the proportion of the <code>k</code> neighbors that are known doublets.
Of the “unmarked” cells that are not known doublets, those with top \(X\) largest proportions are considered to be intra-sample doublets.
We use <code>samples</code> to obtain a reasonable estimate for \(X\), see the vignette for details.</p>
<p>A larger value of <code>k</code> provides more stable estimates of the doublet proportion in each cell.
However, this comes at the cost of assuming that each cell actually has <code>k</code> neighboring cells of the same state.
For example, if a doublet cluster has fewer than <code>k</code> members,
its doublet proportions will be “diluted” by inclusion of unmarked cells in the next-closest cluster.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code>doubletCells</code> and <code>doubletCluster</code>,
for alternative methods of doublet detection when no prior doublet information is available.</p>
<p><code>hashedDrops</code> from the <span class="pkg">DropletUtils</span> package,
to identify doublets from cell hashing experiments.</p>
<p>More detail on the mathematical background of this function is provided in the corresponding vignette at
<code><a href="../articles/recoverDoublets.html">vignette("recoverDoublets", package="scDblFinder")</a></code>.</p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Aaron Lun</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Mocking up an example.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ngenes</span> <span class="op">&lt;-</span> <span class="fl">1000</span></span></span>
<span class="r-in"><span><span class="va">mu1</span> <span class="op">&lt;-</span> <span class="fl">2</span><span class="op">^</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">ngenes</span>, sd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mu2</span> <span class="op">&lt;-</span> <span class="fl">2</span><span class="op">^</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">ngenes</span>, sd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">counts.1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="va">ngenes</span><span class="op">*</span><span class="fl">100</span>, <span class="va">mu1</span><span class="op">)</span>, nrow<span class="op">=</span><span class="va">ngenes</span><span class="op">)</span> <span class="co"># Pure type 1</span></span></span>
<span class="r-in"><span><span class="va">counts.2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="va">ngenes</span><span class="op">*</span><span class="fl">100</span>, <span class="va">mu2</span><span class="op">)</span>, nrow<span class="op">=</span><span class="va">ngenes</span><span class="op">)</span> <span class="co"># Pure type 2</span></span></span>
<span class="r-in"><span><span class="va">counts.m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="va">ngenes</span><span class="op">*</span><span class="fl">20</span>, <span class="va">mu1</span><span class="op">+</span><span class="va">mu2</span><span class="op">)</span>, nrow<span class="op">=</span><span class="va">ngenes</span><span class="op">)</span> <span class="co"># Doublets (1 &amp; 2)</span></span></span>
<span class="r-in"><span><span class="va">all.counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">counts.1</span>, <span class="va">counts.2</span>, <span class="va">counts.m</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">lcounts</span> <span class="op">&lt;-</span> <span class="fu">scuttle</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/normalizeCounts.html" class="external-link">normalizeCounts</a></span><span class="op">(</span><span class="va">all.counts</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Pretending that half of the doublets are known. Also pretending that </span></span></span>
<span class="r-in"><span><span class="co"># the experiment involved two samples of equal size.</span></span></span>
<span class="r-in"><span><span class="va">known</span> <span class="op">&lt;-</span> <span class="fl">200</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span> </span></span>
<span class="r-in"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">recoverDoublets</span><span class="op">(</span><span class="va">lcounts</span>, doublets<span class="op">=</span><span class="va">known</span>, k<span class="op">=</span><span class="fl">10</span>, samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">out</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 220 rows and 3 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     proportion     known predicted</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      &lt;numeric&gt; &lt;logical&gt; &lt;logical&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1            0     FALSE     FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2            0     FALSE     FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3            0     FALSE     FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4            0     FALSE     FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5            0     FALSE     FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ...        ...       ...       ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 216        0.5     FALSE      TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 217        0.5     FALSE      TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 218        0.4     FALSE      TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 219        0.3     FALSE      TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 220        0.4     FALSE      TRUE</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

