<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Scoring potential doublets from simulated densities • scDblFinder</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Scoring potential doublets from simulated densities">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDblFinder</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.21.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/scDblFinder.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/computeDoubletDensity.html">Scoring potential doublets from simulated densities</a></li>
    <li><a class="dropdown-item" href="../articles/findDoubletClusters.html">Detecting clusters of doublet cells with DE analyses</a></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Introduction to the scDblFinder package</a></li>
    <li><a class="dropdown-item" href="../articles/recoverDoublets.html">Recovering intra-sample doublets</a></li>
    <li><a class="dropdown-item" href="../articles/scATAC.html">Doublet identifiation in single-cell ATAC-seq</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/plger/scDblFinder/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Scoring potential doublets from simulated densities</h1>
                        <h4 data-toc-skip class="author">Aaron Lun</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:infinite.monkeys.with.keyboards@gmail.com" class="email">infinite.monkeys.with.keyboards@gmail.com</a>
      
                  
            <h4 data-toc-skip class="date">2025-01-08</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/plger/scDblFinder/blob/devel/vignettes/computeDoubletDensity.Rmd" class="external-link"><code>vignettes/computeDoubletDensity.Rmd</code></a></small>
      <div class="d-none name"><code>computeDoubletDensity.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="tldr">tl;dr<a class="anchor" aria-label="anchor" href="#tldr"></a>
</h2>
<p>To demonstrate, we’ll use one of the mammary gland datasets from the
<em><a href="https://bioconductor.org/packages/3.21/scRNAseq" class="external-link">scRNAseq</a></em>
package. We will subset it down to a random set of 1000 cells for
speed.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/BachMammaryData.html" class="external-link">BachMammaryData</a></span><span class="op">(</span>samples<span class="op">=</span><span class="st">"G_1"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1001</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="fl">1000</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>For the purposes of this demonstration, we’ll perform an extremely
expedited analysis. One would usually take more care here and do some
quality control, create some diagnostic plots, etc., but we don’t have
the space for that.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n<span class="op">=</span><span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1002</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents<span class="op">=</span><span class="fl">10</span>, subset_row<span class="op">=</span><span class="va">hvgs</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">sce</span>, dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></span></code></pre></div>
<p>We run <code><a href="../reference/computeDoubletDensity.html">computeDoubletDensity()</a></code> to obtain a doublet score
for each cell based on the density of simulated doublets around it. We
log this to get some better dynamic range.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1003</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/plger/scDblFinder" class="external-link">scDblFinder</a></span><span class="op">)</span></span>
<span><span class="va">scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeDoubletDensity.html">computeDoubletDensity</a></span><span class="op">(</span><span class="va">sce</span>, subset.row<span class="op">=</span><span class="va">hvgs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/AsIs.html" class="external-link">I</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">scores</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="computeDoubletDensity_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="overview">Algorithm overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>We use a fairly simple approach in <code>doubletCells</code> that
involves creating simulated doublets from the original data set:</p>
<ol style="list-style-type: decimal">
<li>Perform a PCA on the log-normalized expression for all cells in the
dataset.</li>
<li>Randomly select two cells and add their count profiles together.
Compute the log-normalized profile and project it into the PC
space.</li>
<li>Repeat <strong>2</strong> to obtain
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>s</mi></msub><annotation encoding="application/x-tex">N_s</annotation></semantics></math>
simulated doublet cells.</li>
<li>For each cell, compute the local density of simulated doublets,
scaled by the density of the original cells. This is used as the doublet
score.</li>
</ol>
</div>
<div class="section level2">
<h2 id="size-factor-handling">Size factor handling<a class="anchor" aria-label="anchor" href="#size-factor-handling"></a>
</h2>
<div class="section level3">
<h3 id="normalization-size-factors">Normalization size factors<a class="anchor" aria-label="anchor" href="#normalization-size-factors"></a>
</h3>
<p>We allow specification of two sets of size factors for different
purposes. The first set is the normalization set: division of counts by
these size factors yields expression values to be compared across cells.
This is necessary to compute log-normalized expression values for the
PCA.</p>
<p>These size factors are usually computed from some method that assumes
most genes are not DE. We default to library size normalization though
any arbitrary set of size factors can be used. The size factor for each
doublet is computed as the sum of size factors for the individual cells,
based on the additivity of scaling biases.</p>
</div>
<div class="section level3">
<h3 id="rna-content-size-factors">RNA content size factors<a class="anchor" aria-label="anchor" href="#rna-content-size-factors"></a>
</h3>
<p>The second set is the RNA content set: division of counts by these
size factors yields expression values that are proportional to absolute
abundance across cells. This affects the creation of simulated doublets
by controlling the scaling of the count profiles for the individual
cells. These size factors would normally be estimated with spike-ins,
but in their absence we default to using unity for all cells.</p>
<p>The use of unity values implies that the library size for each cell
is a good proxy for total RNA content. This is unlikely to be true:
technical biases mean that the library size is an imprecise relative
estimate of the content. Saturation effects and composition biases also
mean that the expected library size for each population is not an
accurate estimate of content. The imprecision will spread out the
simulated doublets while the inaccuracy will result in a systematic
shift from the location of true doublets.</p>
<p>Arguably, such problems exist for any doublet estimation method
without spike-in information. We can only hope that the inaccuracies
have only minor effects on the creation of simulated cells. Indeed, the
first effect does mitigate the second to some extent by ensuring that
some simulated doublets will occupy the neighbourhood of the true
doublets.</p>
</div>
<div class="section level3">
<h3 id="interactions-between-them">Interactions between them<a class="anchor" aria-label="anchor" href="#interactions-between-them"></a>
</h3>
<p>These two sets of size factors play different roles so it is possible
to specify both of them. We use the following algorithm to accommodate
non-unity values for the RNA content size factors:</p>
<ol style="list-style-type: decimal">
<li>The RNA content size factors are used to scale the counts first.
This ensures that RNA content has the desired effect in step
<strong>2</strong> of Section @ref(overview).</li>
<li>The normalization size factors are also divided by the content size
factors. This ensures that normalization has the correct effect, see
below.</li>
<li>The rest of the algorithm proceeds as if the RNA content size
factors were unity. Addition of count profiles is done without further
scaling, and normalized expression values are computed with the rescaled
normalization size factors.</li>
</ol>
<p>To understand the correctness of the rescaled normalization size
factors, consider a non-DE gene with abundance
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>λ</mi><mi>g</mi></msub><annotation encoding="application/x-tex">\lambda_g</annotation></semantics></math>.
The expected count in each cell is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>λ</mi><mi>g</mi></msub><msub><mi>s</mi><mi>i</mi></msub></mrow><annotation encoding="application/x-tex">\lambda_g s_i</annotation></semantics></math>
for scaling bias
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>s</mi><mi>i</mi></msub><annotation encoding="application/x-tex">s_i</annotation></semantics></math>
(i.e., normalization size factor). The rescaled count is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>λ</mi><mi>g</mi></msub><msub><mi>s</mi><mi>i</mi></msub><msubsup><mi>c</mi><mi>i</mi><mrow><mo>−</mo><mn>1</mn></mrow></msubsup></mrow><annotation encoding="application/x-tex">\lambda_g s_i c_i^{-1}</annotation></semantics></math>
for some RNA content size factor
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>c</mi><mi>i</mi></msub><annotation encoding="application/x-tex">c_i</annotation></semantics></math>.
The rescaled normalization size factor is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>s</mi><mi>i</mi></msub><msubsup><mi>c</mi><mi>i</mi><mrow><mo>−</mo><mn>1</mn></mrow></msubsup></mrow><annotation encoding="application/x-tex">s_i c_i^{-1}</annotation></semantics></math>,
such that normalization yields
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>λ</mi><mi>g</mi></msub><annotation encoding="application/x-tex">\lambda_g</annotation></semantics></math>
as desired. This also holds for doublets where the scaling biases and
size factors are additive.</p>
</div>
</div>
<div class="section level2">
<h2 id="doublet-score-calculations">Doublet score calculations<a class="anchor" aria-label="anchor" href="#doublet-score-calculations"></a>
</h2>
<p>We assume that the simulation accurately mimics doublet creation -
amongst other things, we assume that doublets are equally likely to form
between any cell populations and any differences in total RNA between
subpopulations are captured or negligible. If these assumptions hold,
then at any given region in the expression space, the number of doublets
among the real cells is proportional to the number of simulated doublets
lying in the same region. Thus, the probability that a cell is a doublet
is proportional to the ratio of the number of neighboring simulated
doublets to the number of neighboring real cells.</p>
<p>A mild additional challenge here is that the number of simulated
cells
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>s</mi></msub><annotation encoding="application/x-tex">N_s</annotation></semantics></math>
can vary. Ideally, we would like the expected output of the function to
be the same regardless of the user’s choice of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>s</mi></msub><annotation encoding="application/x-tex">N_s</annotation></semantics></math>,
i.e., the chosen value should only affect the precision/speed trade-off.
Many other doublet-based methods take a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>-nearest
neighbours approach to compute densities; but if
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>s</mi></msub><annotation encoding="application/x-tex">N_s</annotation></semantics></math>
is too large relative to the number of real cells, all of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
nearest neighbours will be simulated, while if
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>s</mi></msub><annotation encoding="application/x-tex">N_s</annotation></semantics></math>
is too small, all of the nearest neighbors will be original cells.</p>
<p>Thus, we use a modified version of the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>NN
approach whereby we identify the distance from each cell to its
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>-th
nearest neighbor. This defines a hypersphere around that cell in which
we count the number of simulated cells. We then compute the odds ratio
of the number of simulated cells in the hypersphere to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>s</mi></msub><annotation encoding="application/x-tex">N_s</annotation></semantics></math>,
divided by the ratio of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
to the total number of cells in the dataset. This score captures the
relative frequency of simulated cells to real cells while being robust
to changes to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mi>s</mi></msub><annotation encoding="application/x-tex">N_s</annotation></semantics></math>.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2025-01-04 r87523)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Etc/UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] bluster_1.17.0              scDblFinder_1.21.2         </span></span>
<span><span class="co">##  [3] scater_1.35.0               ggplot2_3.5.1              </span></span>
<span><span class="co">##  [5] scran_1.35.0                scuttle_1.17.0             </span></span>
<span><span class="co">##  [7] ensembldb_2.31.0            AnnotationFilter_1.31.0    </span></span>
<span><span class="co">##  [9] GenomicFeatures_1.59.1      AnnotationDbi_1.69.0       </span></span>
<span><span class="co">## [11] scRNAseq_2.21.0             SingleCellExperiment_1.29.1</span></span>
<span><span class="co">## [13] SummarizedExperiment_1.37.0 Biobase_2.67.0             </span></span>
<span><span class="co">## [15] GenomicRanges_1.59.1        GenomeInfoDb_1.43.2        </span></span>
<span><span class="co">## [17] IRanges_2.41.2              S4Vectors_0.45.2           </span></span>
<span><span class="co">## [19] BiocGenerics_0.53.3         generics_0.1.3             </span></span>
<span><span class="co">## [21] MatrixGenerics_1.19.0       matrixStats_1.5.0          </span></span>
<span><span class="co">## [23] BiocStyle_2.35.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] jsonlite_1.8.9           magrittr_2.0.3           ggbeeswarm_0.7.2        </span></span>
<span><span class="co">##   [4] gypsum_1.3.0             farver_2.1.2             rmarkdown_2.29          </span></span>
<span><span class="co">##   [7] fs_1.6.5                 BiocIO_1.17.1            zlibbioc_1.53.0         </span></span>
<span><span class="co">##  [10] ragg_1.3.3               vctrs_0.6.5              memoise_2.0.1           </span></span>
<span><span class="co">##  [13] Rsamtools_2.23.1         RCurl_1.98-1.16          htmltools_0.5.8.1       </span></span>
<span><span class="co">##  [16] S4Arrays_1.7.1           AnnotationHub_3.15.0     curl_6.1.0              </span></span>
<span><span class="co">##  [19] BiocNeighbors_2.1.2      xgboost_1.7.8.1          Rhdf5lib_1.29.0         </span></span>
<span><span class="co">##  [22] SparseArray_1.7.2        rhdf5_2.51.1             sass_0.4.9              </span></span>
<span><span class="co">##  [25] alabaster.base_1.7.2     bslib_0.8.0              htmlwidgets_1.6.4       </span></span>
<span><span class="co">##  [28] desc_1.4.3               alabaster.sce_1.7.0      httr2_1.0.7             </span></span>
<span><span class="co">##  [31] cachem_1.1.0             GenomicAlignments_1.43.0 igraph_2.1.3            </span></span>
<span><span class="co">##  [34] mime_0.12                lifecycle_1.0.4          pkgconfig_2.0.3         </span></span>
<span><span class="co">##  [37] rsvd_1.0.5               Matrix_1.7-1             R6_2.5.1                </span></span>
<span><span class="co">##  [40] fastmap_1.2.0            GenomeInfoDbData_1.2.13  digest_0.6.37           </span></span>
<span><span class="co">##  [43] colorspace_2.1-1         dqrng_0.4.1              irlba_2.3.5.1           </span></span>
<span><span class="co">##  [46] ExperimentHub_2.15.0     textshaping_0.4.1        RSQLite_2.3.9           </span></span>
<span><span class="co">##  [49] beachmat_2.23.6          labeling_0.4.3           filelock_1.0.3          </span></span>
<span><span class="co">##  [52] httr_1.4.7               abind_1.4-8              compiler_4.5.0          </span></span>
<span><span class="co">##  [55] bit64_4.5.2              withr_3.0.2              BiocParallel_1.41.0     </span></span>
<span><span class="co">##  [58] viridis_0.6.5            DBI_1.2.3                HDF5Array_1.35.2        </span></span>
<span><span class="co">##  [61] alabaster.ranges_1.7.0   alabaster.schemas_1.7.0  MASS_7.3-64             </span></span>
<span><span class="co">##  [64] rappdirs_0.3.3           DelayedArray_0.33.3      rjson_0.2.23            </span></span>
<span><span class="co">##  [67] tools_4.5.0              vipor_0.4.7              beeswarm_0.4.0          </span></span>
<span><span class="co">##  [70] glue_1.8.0               restfulr_0.0.15          rhdf5filters_1.19.0     </span></span>
<span><span class="co">##  [73] grid_4.5.0               Rtsne_0.17               cluster_2.1.8           </span></span>
<span><span class="co">##  [76] gtable_0.3.6             data.table_1.16.4        metapod_1.15.0          </span></span>
<span><span class="co">##  [79] BiocSingular_1.23.0      ScaledMatrix_1.15.0      XVector_0.47.1          </span></span>
<span><span class="co">##  [82] ggrepel_0.9.6            BiocVersion_3.21.1       pillar_1.10.1           </span></span>
<span><span class="co">##  [85] limma_3.63.2             dplyr_1.1.4              BiocFileCache_2.15.0    </span></span>
<span><span class="co">##  [88] lattice_0.22-6           rtracklayer_1.67.0       bit_4.5.0.1             </span></span>
<span><span class="co">##  [91] tidyselect_1.2.1         locfit_1.5-9.10          Biostrings_2.75.3       </span></span>
<span><span class="co">##  [94] knitr_1.49               gridExtra_2.3            bookdown_0.42           </span></span>
<span><span class="co">##  [97] ProtGenerics_1.39.1      edgeR_4.5.1              xfun_0.50               </span></span>
<span><span class="co">## [100] statmod_1.5.0            UCSC.utils_1.3.0         lazyeval_0.2.2          </span></span>
<span><span class="co">## [103] yaml_2.3.10              evaluate_1.0.1           codetools_0.2-20        </span></span>
<span><span class="co">## [106] tibble_3.2.1             alabaster.matrix_1.7.4   BiocManager_1.30.25     </span></span>
<span><span class="co">## [109] cli_3.6.3                systemfonts_1.1.0        munsell_0.5.1           </span></span>
<span><span class="co">## [112] jquerylib_0.1.4          Rcpp_1.0.13-1            dbplyr_2.5.0            </span></span>
<span><span class="co">## [115] png_0.1-8                XML_3.99-0.18            parallel_4.5.0          </span></span>
<span><span class="co">## [118] pkgdown_2.1.1            blob_1.2.4               bitops_1.0-9            </span></span>
<span><span class="co">## [121] viridisLite_0.4.2        alabaster.se_1.7.0       scales_1.3.0            </span></span>
<span><span class="co">## [124] purrr_1.0.2              crayon_1.5.3             rlang_1.1.4             </span></span>
<span><span class="co">## [127] KEGGREST_1.47.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
