<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Detecting clusters of doublet cells with DE analyses • scDblFinder</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Detecting clusters of doublet cells with DE analyses">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDblFinder</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.21.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/scDblFinder.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/computeDoubletDensity.html">Scoring potential doublets from simulated densities</a></li>
    <li><a class="dropdown-item" href="../articles/findDoubletClusters.html">Detecting clusters of doublet cells with DE analyses</a></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Introduction to the scDblFinder package</a></li>
    <li><a class="dropdown-item" href="../articles/recoverDoublets.html">Recovering intra-sample doublets</a></li>
    <li><a class="dropdown-item" href="../articles/scATAC.html">Doublet identifiation in single-cell ATAC-seq</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/plger/scDblFinder/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Detecting clusters of doublet cells with DE analyses</h1>
                        <h4 data-toc-skip class="author">Aaron Lun</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:infinite.monkeys.with.keyboards@gmail.com" class="email">infinite.monkeys.with.keyboards@gmail.com</a>
      
                  
            <h4 data-toc-skip class="date">2025-01-08</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/plger/scDblFinder/blob/devel/vignettes/findDoubletClusters.Rmd" class="external-link"><code>vignettes/findDoubletClusters.Rmd</code></a></small>
      <div class="d-none name"><code>findDoubletClusters.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="tldr">tl;dr<a class="anchor" aria-label="anchor" href="#tldr"></a>
</h2>
<p>To demonstrate, we’ll use one of the mammary gland datasets from the
<em><a href="https://bioconductor.org/packages/3.21/scRNAseq" class="external-link">scRNAseq</a></em>
package. We will subset it down to a random set of 500 cells for
speed.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/BachMammaryData.html" class="external-link">BachMammaryData</a></span><span class="op">(</span>samples<span class="op">=</span><span class="st">"G_2"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="fl">500</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>For the purposes of this demonstration, we’ll perform an extremely
expedited analysis. One would usually take more care here and do some
quality control, create some diagnostic plots, etc., but we don’t have
the space for that.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents<span class="op">=</span><span class="fl">10</span>, subset_row<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n<span class="op">=</span><span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/clusterRows.html" class="external-link">clusterRows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCA"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">NNGraphParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">sce</span>, dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/AsIs.html" class="external-link">I</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span>, text_by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/AsIs.html" class="external-link">I</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="findDoubletClusters_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>We then run <code><a href="../reference/findDoubletClusters.html">findDoubletClusters()</a></code> to test each cluster
against the null hypothesis that it <em>does</em> consist of doublets.
The null is rejected if a cluster has many DE genes that lie outside the
expression limits defined by the “source” clusters. On the other hand,
if <code>num.de</code> is low, the cluster’s expression profile is
consistent with the doublet hypothesis.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/plger/scDblFinder" class="external-link">scDblFinder</a></span><span class="op">)</span></span>
<span><span class="va">tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findDoubletClusters.html">findDoubletClusters</a></span><span class="op">(</span><span class="va">sce</span>, <span class="va">clusters</span><span class="op">)</span></span>
<span><span class="va">tab</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 7 rows and 9 columns</span></span>
<span><span class="co">##       source1     source2    num.de median.de               best     p.value</span></span>
<span><span class="co">##   &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt;        &lt;character&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">## 6           2           1         0       122 ENSMUSG00000019256 1.00000e+00</span></span>
<span><span class="co">## 3           6           5         8        46 ENSMUSG00000002985 8.41352e-05</span></span>
<span><span class="co">## 7           5           1        11       393 ENSMUSG00000075705 2.04690e-10</span></span>
<span><span class="co">## 4           7           2        24        63 ENSMUSG00000022491 4.15645e-13</span></span>
<span><span class="co">## 1           7           6        86       715 ENSMUSG00000001349 8.49703e-08</span></span>
<span><span class="co">## 5           7           2       112      1007 ENSMUSG00000024610 2.38765e-11</span></span>
<span><span class="co">## 2           6           5       124       336 ENSMUSG00000023505 4.06936e-06</span></span>
<span><span class="co">##   lib.size1 lib.size2      prop</span></span>
<span><span class="co">##   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 6  0.927081  0.443667     0.084</span></span>
<span><span class="co">## 3  2.026090  0.859424     0.242</span></span>
<span><span class="co">## 7  1.195040  1.249946     0.030</span></span>
<span><span class="co">## 4  0.630479  1.646729     0.150</span></span>
<span><span class="co">## 1  0.800035  2.253942     0.192</span></span>
<span><span class="co">## 5  0.836792  2.185591     0.032</span></span>
<span><span class="co">## 2  1.078655  0.457542     0.270</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="mathematical-background">Mathematical background<a class="anchor" aria-label="anchor" href="#mathematical-background"></a>
</h2>
<p>Consider a cell population
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
that has mean transcript count
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>λ</mi><mrow><mi>g</mi><mi>i</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{gi}</annotation></semantics></math>
for gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>.
Assume that each population exhibits a unique scaling bias
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>s</mi><mi>i</mi></msub><annotation encoding="application/x-tex">s_i</annotation></semantics></math>,
representing the efficiency of library preparation for that population.
The observed read/UMI count for each gene is then
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>μ</mi><mrow><mi>g</mi><mi>i</mi></mrow></msub><mo>=</mo><msub><mi>s</mi><mi>i</mi></msub><msub><mi>λ</mi><mrow><mi>g</mi><mi>i</mi></mrow></msub></mrow><annotation encoding="application/x-tex">\mu_{gi}=s_i\lambda_{gi}</annotation></semantics></math>.
(For simplicity, we will ignore gene-specific scaling biases, as this is
easily accommodated by considering
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>λ</mi><mrow><mi>g</mi><mi>i</mi></mrow></msub><mo>≡</mo><msub><mi>ϕ</mi><mi>g</mi></msub><msub><mi>λ</mi><mrow><mi>g</mi><mi>i</mi></mrow></msub></mrow><annotation encoding="application/x-tex">\lambda_{gi} \equiv \phi_g \lambda_{gi}</annotation></semantics></math>
for some bias
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ϕ</mi><mi>g</mi></msub><annotation encoding="application/x-tex">\phi_g</annotation></semantics></math>.)
The expected total count for each population is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>N</mi><mi>i</mi></msub><mo>=</mo><msub><mo>∑</mo><mi>g</mi></msub><msub><mi>μ</mi><mrow><mi>g</mi><mi>i</mi></mrow></msub></mrow><annotation encoding="application/x-tex">N_i = \sum_g \mu_{gi}</annotation></semantics></math>.</p>
<p>Now, let us consider a doublet population
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
that forms from two parent populations
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>i</mi><mn>1</mn></msub><annotation encoding="application/x-tex">i_1</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>i</mi><mn>2</mn></msub><annotation encoding="application/x-tex">i_2</annotation></semantics></math>.
The observed read count for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>μ</mi><mrow><mi>g</mi><mi>j</mi></mrow></msub><mo>=</mo><msub><mi>s</mi><mi>j</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>1</mn></msub></mrow></msub><mo>+</mo><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>2</mn></msub></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\mu_{gj} = s_j (\lambda_{gi_1} + \lambda_{gi_2})</annotation></semantics></math>.
Note that
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>s</mi><mi>j</mi></msub><annotation encoding="application/x-tex">s_j</annotation></semantics></math>
need not be any particular function of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>s</mi><msub><mi>i</mi><mn>1</mn></msub></msub><annotation encoding="application/x-tex">s_{i_1}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>s</mi><msub><mi>i</mi><mn>2</mn></msub></msub><annotation encoding="application/x-tex">s_{i_2}</annotation></semantics></math>.
Rather, this relationship depends on how quickly the reverse
transcription and amplification reagents are saturated during library
preparation, which is difficult to make assumptions around.</p>
</div>
<div class="section level2">
<h2 id="normalization-by-library-size">Normalization by library size<a class="anchor" aria-label="anchor" href="#normalization-by-library-size"></a>
</h2>
<p>We obtain log-normalized expression values for each cell based on the
library size. Assume that the library size-normalized expression values
are such that
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>μ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>1</mn></msub></mrow></msub><msubsup><mi>N</mi><msub><mi>i</mi><mn>1</mn></msub><mrow><mo>−</mo><mn>1</mn></mrow></msubsup><mo>&lt;</mo><msub><mi>μ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>2</mn></msub></mrow></msub><msubsup><mi>N</mi><msub><mi>i</mi><mn>2</mn></msub><mrow><mo>−</mo><mn>1</mn></mrow></msubsup></mrow><annotation encoding="application/x-tex">\mu_{gi_1}N_{i_1}^{-1} &lt; \mu_{gi_2}N_{i_2}^{-1}</annotation></semantics></math>,
i.e., the proportion of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
increases in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>i</mi><mn>2</mn></msub><annotation encoding="application/x-tex">i_2</annotation></semantics></math>
compared to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>i</mi><mn>1</mn></msub><annotation encoding="application/x-tex">i_1</annotation></semantics></math>.
The contribution of each
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>s</mi><mi>i</mi></msub><annotation encoding="application/x-tex">s_i</annotation></semantics></math>
cancels out, yielding
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>1</mn></msub></mrow></msub><mrow><munder><mo>∑</mo><mi>g</mi></munder><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>1</mn></msub></mrow></msub></mrow></mfrac><mo>&lt;</mo><mfrac><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>2</mn></msub></mrow></msub><mrow><munder><mo>∑</mo><mi>g</mi></munder><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>2</mn></msub></mrow></msub></mrow></mfrac><mspace width="0.278em"></mspace><mi>.</mi></mrow><annotation encoding="application/x-tex">
\frac{\lambda_{gi_1}}{\sum_g \lambda_{gi_1}} &lt; \frac{\lambda_{gi_2}}{\sum_g \lambda_{gi_2}} \;.
</annotation></semantics></math> The normalized expression value of the
doublet cluster
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
is subsequently
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><mrow><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>1</mn></msub></mrow></msub><mo>+</mo><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>2</mn></msub></mrow></msub></mrow><mrow><munder><mo>∑</mo><mi>g</mi></munder><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>1</mn></msub></mrow></msub><mo>+</mo><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>2</mn></msub></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mfrac><mspace width="0.278em"></mspace><mo>,</mo></mrow><annotation encoding="application/x-tex">
\frac{\lambda_{gi_1} + \lambda_{gi_2}}{\sum_g (\lambda_{gi_1} + \lambda_{gi_2})} \;,
</annotation></semantics></math> and it is fairly easy to show that
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>1</mn></msub></mrow></msub><mrow><munder><mo>∑</mo><mi>g</mi></munder><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>1</mn></msub></mrow></msub></mrow></mfrac><mo>&lt;</mo><mfrac><mrow><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>1</mn></msub></mrow></msub><mo>+</mo><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>2</mn></msub></mrow></msub></mrow><mrow><munder><mo>∑</mo><mi>g</mi></munder><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>1</mn></msub></mrow></msub><mo>+</mo><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>2</mn></msub></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mfrac><mo>&lt;</mo><mfrac><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>2</mn></msub></mrow></msub><mrow><munder><mo>∑</mo><mi>g</mi></munder><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>2</mn></msub></mrow></msub></mrow></mfrac><mspace width="0.278em"></mspace><mi>.</mi></mrow><annotation encoding="application/x-tex">
\frac{\lambda_{gi_1}}{\sum_g \lambda_{gi_1}} &lt; 
\frac{\lambda_{gi_1} + \lambda_{gi_2}}{\sum_g (\lambda_{gi_1} + \lambda_{gi_2})}  &lt; 
\frac{\lambda_{gi_2}}{\sum_g \lambda_{gi_2}} \;.
</annotation></semantics></math> In other words, the expected library
size-normalized expression of our gene in the doublet cluster lies
between that of the two parents.</p>
<p>It is harder to provide theoretical guarantees with arbitrary size
factors, which is why we only use the library sizes for normalization
instead. The exception is that of spike-in size factors that would
estimate
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>s</mi><mi>i</mi></msub><annotation encoding="application/x-tex">s_i</annotation></semantics></math>
directly. This would allow us to obtain estimates of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>λ</mi><mrow><mi>g</mi><mi>i</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{gi}</annotation></semantics></math>
for the parent clusters and of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>1</mn></msub></mrow></msub><mo>+</mo><msub><mi>λ</mi><mrow><mi>g</mi><msub><mi>i</mi><mn>2</mn></msub></mrow></msub></mrow><annotation encoding="application/x-tex">\lambda_{gi_1} + \lambda_{gi_2}</annotation></semantics></math>
for the doublets. In this manner, we could more precisely identify
doublet clusters as those where the normalized expression value is equal
to the sum of the parents. Unfortunately, spike-ins are generally not
available for droplet-based data sets where doublets are most
problematic.</p>
</div>
<div class="section level2">
<h2 id="testing-for-lack-of-intermediacy">Testing for (lack of) intermediacy<a class="anchor" aria-label="anchor" href="#testing-for-lack-of-intermediacy"></a>
</h2>
<p>We want to identify the clusters that may be comprised of doublets of
other clusters. For each cluster
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>j</mi><mi>′</mi></mrow><annotation encoding="application/x-tex">j'</annotation></semantics></math>,
we test for differential expression in the library size-normalized
expression profiles against every other cluster
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><mi>′</mi></mrow><annotation encoding="application/x-tex">i'</annotation></semantics></math>.
For each pair of other clusters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><msub><mi>′</mi><mn>1</mn></msub></mrow><annotation encoding="application/x-tex">i'_1</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><msub><mi>′</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">i'_2</annotation></semantics></math>,
we identify genes that change in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>j</mi><mi>′</mi></mrow><annotation encoding="application/x-tex">j'</annotation></semantics></math>
against both
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><msub><mi>′</mi><mn>1</mn></msub></mrow><annotation encoding="application/x-tex">i'_1</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><msub><mi>′</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">i'_2</annotation></semantics></math><strong>in the same direction</strong>. The presence of such genes
violates the intermediacy expected of a doublet cluster and provides
evidence that
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>j</mi><mi>′</mi></mrow><annotation encoding="application/x-tex">j'</annotation></semantics></math>
is not a doublet of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><msub><mi>′</mi><mn>1</mn></msub></mrow><annotation encoding="application/x-tex">i'_1</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><msub><mi>′</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">i'_2</annotation></semantics></math>.</p>
<p>Significant genes are identified by an intersection-union test on the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values
from the pairwise comparisons between
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>j</mi><mi>′</mi></mrow><annotation encoding="application/x-tex">j'</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><msub><mi>′</mi><mn>1</mn></msub></mrow><annotation encoding="application/x-tex">i'_1</annotation></semantics></math>
or
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>i</mi><msub><mi>′</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">i'_2</annotation></semantics></math>.
(Specifically,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>-tests
are used via the <code><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers()</a></code> function from <em><a href="https://bioconductor.org/packages/3.21/scran" class="external-link">scran</a></em>.) The
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-value
for a gene is set to unity when the signs of the log-fold changes are
not the same between comparisons. Multiple correction testing is applied
using the Benjamini-Hochberg method, and the number of genes detected at
a specified false discovery rate (usually 5%) is counted. The pair
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><msub><mi>′</mi><mn>1</mn></msub><mo>,</mo><mi>i</mi><msub><mi>′</mi><mn>2</mn></msub><mo stretchy="true" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">(i'_1, i'_2)</annotation></semantics></math>
with the fewest detected genes are considered as the putative parents of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>j</mi><mi>′</mi></mrow><annotation encoding="application/x-tex">j'</annotation></semantics></math>.</p>
<p>In theory, it is possible to compute the Simes’ combined
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-value
across all genes to reject the doublet hypothesis for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>j</mi><mi>′</mi></mrow><annotation encoding="application/x-tex">j'</annotation></semantics></math>.
This would provide a more rigorous approach to ruling out potential
doublet/parent combinations. However, this is very sensitive to
misspecification of clusters – see below.</p>
</div>
<div class="section level2">
<h2 id="calling-doublet-clusters">Calling doublet clusters<a class="anchor" aria-label="anchor" href="#calling-doublet-clusters"></a>
</h2>
<p>Assuming that most clusters are not comprised of doublets, we
identify clusters that have an unusually low number of detected genes
that violate the intermediacy condition. This is achieved by identifying
small outliers on the log-transformed number of detected genes, using
the median absolute deviation-based method in the function. (We use a
log-transformation simply to improve resolution at low values.) Clusters
are likely to be doublets if they are outliers on this metric.</p>
<p>Doublet clusters should also have larger library sizes than the
proposed parent clusters. This is consistent with the presence of more
RNA in each doublet, though the library size of the doublet cluster need
not be a sum of that of the parent clusters (due to factors such as
saturation and composition effects). The proportion of cells assigned to
the doublet cluster should also be “reasonable”; exactly what this means
depends on the experimental setup and the doublet rate of the protocol
in use.</p>
</div>
<div class="section level2">
<h2 id="discussion">Discussion<a class="anchor" aria-label="anchor" href="#discussion"></a>
</h2>
<p>The biggest advantage of this approach lies in its interpretability.
Given a set of existing clusters, we can explicitly identify those that
are likely to be doublets. We also gain some insight onto the parental
origins of each putative doublet cluster, which may be of some interest.
We avoid any assumptions about doublet formation that are otherwise
necessary for the simulation-based methods. In particular, we do not
require any knowledge about exact the relationship between
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>s</mi><mi>j</mi></msub><annotation encoding="application/x-tex">s_j</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>s</mi><mi>i</mi></msub><annotation encoding="application/x-tex">s_i</annotation></semantics></math>,
allowing us to identify doublets even when the exact location of the
doublet is unknown (e.g., due to differences in RNA content between the
parent clusters).</p>
<p>The downside is that, of course, we are dependent on being supplied
with sensible clusters where the parental and doublet cells are
separated. The intermediacy requirement is loose enough to provide some
robustness against misspecification, but this only goes so far. In
addition, this strategy has a bias towards calling clusters with few
cells as doublets (or parents of doublets) because the DE detection
power is low. This can be somewhat offset by comparing
<code>num.de</code> against <code>median.de</code> as latter will be low
for clusters involved in systematically low-powered comparisons, though
it is difficult to adjust for the exact effect of the differences of
power on the IUT.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2025-01-04 r87523)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Etc/UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] scDblFinder_1.21.2          bluster_1.17.0             </span></span>
<span><span class="co">##  [3] scater_1.35.0               ggplot2_3.5.1              </span></span>
<span><span class="co">##  [5] scran_1.35.0                scuttle_1.17.0             </span></span>
<span><span class="co">##  [7] ensembldb_2.31.0            AnnotationFilter_1.31.0    </span></span>
<span><span class="co">##  [9] GenomicFeatures_1.59.1      AnnotationDbi_1.69.0       </span></span>
<span><span class="co">## [11] scRNAseq_2.21.0             SingleCellExperiment_1.29.1</span></span>
<span><span class="co">## [13] SummarizedExperiment_1.37.0 Biobase_2.67.0             </span></span>
<span><span class="co">## [15] GenomicRanges_1.59.1        GenomeInfoDb_1.43.2        </span></span>
<span><span class="co">## [17] IRanges_2.41.2              S4Vectors_0.45.2           </span></span>
<span><span class="co">## [19] BiocGenerics_0.53.3         generics_0.1.3             </span></span>
<span><span class="co">## [21] MatrixGenerics_1.19.0       matrixStats_1.5.0          </span></span>
<span><span class="co">## [23] BiocStyle_2.35.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] jsonlite_1.8.9           magrittr_2.0.3           ggbeeswarm_0.7.2        </span></span>
<span><span class="co">##   [4] gypsum_1.3.0             farver_2.1.2             rmarkdown_2.29          </span></span>
<span><span class="co">##   [7] fs_1.6.5                 BiocIO_1.17.1            zlibbioc_1.53.0         </span></span>
<span><span class="co">##  [10] ragg_1.3.3               vctrs_0.6.5              memoise_2.0.1           </span></span>
<span><span class="co">##  [13] Rsamtools_2.23.1         RCurl_1.98-1.16          htmltools_0.5.8.1       </span></span>
<span><span class="co">##  [16] S4Arrays_1.7.1           AnnotationHub_3.15.0     curl_6.1.0              </span></span>
<span><span class="co">##  [19] BiocNeighbors_2.1.2      xgboost_1.7.8.1          Rhdf5lib_1.29.0         </span></span>
<span><span class="co">##  [22] SparseArray_1.7.2        rhdf5_2.51.1             sass_0.4.9              </span></span>
<span><span class="co">##  [25] alabaster.base_1.7.2     bslib_0.8.0              htmlwidgets_1.6.4       </span></span>
<span><span class="co">##  [28] desc_1.4.3               alabaster.sce_1.7.0      httr2_1.0.7             </span></span>
<span><span class="co">##  [31] cachem_1.1.0             GenomicAlignments_1.43.0 igraph_2.1.3            </span></span>
<span><span class="co">##  [34] mime_0.12                lifecycle_1.0.4          pkgconfig_2.0.3         </span></span>
<span><span class="co">##  [37] rsvd_1.0.5               Matrix_1.7-1             R6_2.5.1                </span></span>
<span><span class="co">##  [40] fastmap_1.2.0            GenomeInfoDbData_1.2.13  digest_0.6.37           </span></span>
<span><span class="co">##  [43] colorspace_2.1-1         dqrng_0.4.1              irlba_2.3.5.1           </span></span>
<span><span class="co">##  [46] ExperimentHub_2.15.0     textshaping_0.4.1        RSQLite_2.3.9           </span></span>
<span><span class="co">##  [49] beachmat_2.23.6          labeling_0.4.3           filelock_1.0.3          </span></span>
<span><span class="co">##  [52] httr_1.4.7               abind_1.4-8              compiler_4.5.0          </span></span>
<span><span class="co">##  [55] bit64_4.5.2              withr_3.0.2              BiocParallel_1.41.0     </span></span>
<span><span class="co">##  [58] viridis_0.6.5            DBI_1.2.3                HDF5Array_1.35.2        </span></span>
<span><span class="co">##  [61] alabaster.ranges_1.7.0   alabaster.schemas_1.7.0  MASS_7.3-64             </span></span>
<span><span class="co">##  [64] rappdirs_0.3.3           DelayedArray_0.33.3      rjson_0.2.23            </span></span>
<span><span class="co">##  [67] tools_4.5.0              vipor_0.4.7              beeswarm_0.4.0          </span></span>
<span><span class="co">##  [70] glue_1.8.0               restfulr_0.0.15          rhdf5filters_1.19.0     </span></span>
<span><span class="co">##  [73] grid_4.5.0               Rtsne_0.17               cluster_2.1.8           </span></span>
<span><span class="co">##  [76] gtable_0.3.6             data.table_1.16.4        metapod_1.15.0          </span></span>
<span><span class="co">##  [79] BiocSingular_1.23.0      ScaledMatrix_1.15.0      XVector_0.47.1          </span></span>
<span><span class="co">##  [82] ggrepel_0.9.6            BiocVersion_3.21.1       pillar_1.10.1           </span></span>
<span><span class="co">##  [85] limma_3.63.2             dplyr_1.1.4              BiocFileCache_2.15.0    </span></span>
<span><span class="co">##  [88] lattice_0.22-6           rtracklayer_1.67.0       bit_4.5.0.1             </span></span>
<span><span class="co">##  [91] tidyselect_1.2.1         locfit_1.5-9.10          Biostrings_2.75.3       </span></span>
<span><span class="co">##  [94] knitr_1.49               gridExtra_2.3            bookdown_0.42           </span></span>
<span><span class="co">##  [97] ProtGenerics_1.39.1      edgeR_4.5.1              xfun_0.50               </span></span>
<span><span class="co">## [100] statmod_1.5.0            UCSC.utils_1.3.0         lazyeval_0.2.2          </span></span>
<span><span class="co">## [103] yaml_2.3.10              evaluate_1.0.1           codetools_0.2-20        </span></span>
<span><span class="co">## [106] tibble_3.2.1             alabaster.matrix_1.7.4   BiocManager_1.30.25     </span></span>
<span><span class="co">## [109] cli_3.6.3                systemfonts_1.1.0        munsell_0.5.1           </span></span>
<span><span class="co">## [112] jquerylib_0.1.4          Rcpp_1.0.13-1            dbplyr_2.5.0            </span></span>
<span><span class="co">## [115] png_0.1-8                XML_3.99-0.18            parallel_4.5.0          </span></span>
<span><span class="co">## [118] pkgdown_2.1.1            blob_1.2.4               bitops_1.0-9            </span></span>
<span><span class="co">## [121] viridisLite_0.4.2        alabaster.se_1.7.0       scales_1.3.0            </span></span>
<span><span class="co">## [124] purrr_1.0.2              crayon_1.5.3             rlang_1.1.4             </span></span>
<span><span class="co">## [127] KEGGREST_1.47.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
