<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to the scDblFinder package • scDblFinder</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to the scDblFinder package">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDblFinder</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.19.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/scDblFinder.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/computeDoubletDensity.html">Scoring potential doublets from simulated densities</a></li>
    <li><a class="dropdown-item" href="../articles/findDoubletClusters.html">Detecting clusters of doublet cells with DE analyses</a></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Introduction to the scDblFinder package</a></li>
    <li><a class="dropdown-item" href="../articles/recoverDoublets.html">Recovering intra-sample doublets</a></li>
    <li><a class="dropdown-item" href="../articles/scATAC.html">Doublet identifiation in single-cell ATAC-seq</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/plger/scDblFinder/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Introduction to the scDblFinder package</h1>
                        <h4 data-toc-skip class="author">Pierre-Luc
Germain</h4>
            <address class="author_afil">
      University and ETH
Zürich<br><a class="author_email" href="mailto:#"></a><a href="mailto:pierre-luc.germain@hest.ethz.ch" class="email">pierre-luc.germain@hest.ethz.ch</a>
      </address>
                              <h4 data-toc-skip class="author">Aaron
Lun</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:infinite.monkeys.with.keyboards@gmail.com" class="email">infinite.monkeys.with.keyboards@gmail.com</a>
      
                  
      
      <small class="dont-index">Source: <a href="https://github.com/plger/scDblFinder/blob/devel/vignettes/introduction.Rmd" class="external-link"><code>vignettes/introduction.Rmd</code></a></small>
      <div class="d-none name"><code>introduction.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>An introduction to the various methods included in the
      scDblFinder package.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <code>scDblFinder</code> package gathers various methods for the
detection and handling of doublets/multiplets in single-cell sequencing
data (i.e. multiple cells captured within the same droplet or reaction
volume). This vignette provides a brief overview of the different
approaches (which are each covered in their own vignettes) for
single-cell RNA sequencing. <em>For doublet detection in genomic data,
see the <a href="scATAC.html">scATACseq vignette</a></em>. For a more
general introduction to the topic of doublets, refer to the <a href="https://osca.bioconductor.org/doublet-detection.html" class="external-link">OCSA
book</a>.</p>
<p>All methods require as an input either a matrix of counts or a <em><a href="https://bioconductor.org/packages/3.19/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
containing count data. With the exception of <a href="findDoubletClusters.html">findDoubletClusters</a>, which operates
at the level of clusters (and consequently requires clustering
information), all methods try to assign each cell a score indicating its
likelihood (broadly understood) of being a doublet.</p>
<p>The approaches described here are <em>complementary</em> to doublets
identified via cell hashes and SNPs in multiplexed samples: while
hashing/genotypes can identify doublets formed by cells of the same type
(homotypic doublets) from two samples, which are often nearly
undistinguishable from real cells transcriptionally (and hence generally
unidentifiable through the present package), it cannot identify doublets
made by cells of the same sample, even if they are heterotypic (formed
by different cell types). Instead, the methods presented here are
primarily geared towards the identification of heterotypic doublets,
which for most purposes are also the most critical ones.</p>
<p><br></p>
<div class="section level3">
<h3 id="computedoubletdensity">computeDoubletDensity<a class="anchor" aria-label="anchor" href="#computedoubletdensity"></a>
</h3>
<p>The <code>computeDoubletDensity</code> method (formerly
<code><a href="https://rdrr.io/pkg/scran/man/defunct.html" class="external-link">scran::doubletCells</a></code>) generates random artificial doublets
from the real cells, and tries to identify cells whose neighborhood has
a high local density of articial doublets. See <a href="computeDoubletDensity.html">computeDoubletDensity</a> for more
information.</p>
</div>
<div class="section level3">
<h3 id="recoverdoublets">recoverDoublets<a class="anchor" aria-label="anchor" href="#recoverdoublets"></a>
</h3>
<p>The <code>recoverDoublets</code> method is meant to be used when some
doublets are already known, for instance through genotype-based calls or
cell hashing in multiplexed experiments. The function then tries to
identify intra-sample doublets that are neighbors to the known
inter-sample doublets. See <a href="recoverDoublets.html">recoverDoublets</a> for more
information.</p>
</div>
<div class="section level3">
<h3 id="scdblfinder">scDblFinder<a class="anchor" aria-label="anchor" href="#scdblfinder"></a>
</h3>
<p>The <code>scDblFinder</code> method combines both known doublets (if
available) and cluster-based artificial doublets to identify doublets.
The approach builds and improves on a variety of earlier efforts, and is
at present the most accurate approach included in this package. See <a href="scDblFinder.html">scDblFinder</a> for more information.</p>
</div>
<div class="section level3">
<h3 id="directdblclassification">directDblClassification<a class="anchor" aria-label="anchor" href="#directdblclassification"></a>
</h3>
<p>The <code>directDblClassification</code> method identifies doublets
by training a classifier directly on gene expression. This follows the
same procedure as <code>scDblFinder</code> for doublet generation and
iterative training, but skips the <em>k</em>-nearest neighbor step and
directly uses the matrix of real cells and artificial doublets. This is
computationally more intensive and generally leads to worse predictions
than <code>scDblFinder</code>, and it is included chiefly for
comparative purposes. See <code><a href="../reference/directDblClassification.html">?directDblClassification</a></code> for more
information.</p>
</div>
<div class="section level3">
<h3 id="finddoubletclusters">findDoubletClusters<a class="anchor" aria-label="anchor" href="#finddoubletclusters"></a>
</h3>
<p>The <code>findDoubletClusters</code> method identifies clusters that
are likely to be composed of doublets by estimating whether their
expression profile lies between two other clusters. See <a href="findDoubletClusters.html">findDoubletClusters</a> for more
information.</p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scDblFinder"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># or, to get that latest developments:</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"plger/scDblFinder"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="which-method-to-choose">Which method to choose?<a class="anchor" aria-label="anchor" href="#which-method-to-choose"></a>
</h2>
<p>A benchmark of the main methods available in the package is presented
in the <a href="https://f1000research.com/articles/10-979/" class="external-link">scDblFinder
paper</a>. While the different methods included here have their values,
overall the <code>scDblFinder</code> method had the best performance
(also superior to other methods not included in this package), and
should be used by default.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] BiocStyle_2.32.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] digest_0.6.37       desc_1.4.3          R6_2.5.1           </span></span>
<span><span class="co">##  [4] bookdown_0.40       fastmap_1.2.0       xfun_0.47          </span></span>
<span><span class="co">##  [7] cachem_1.1.0        knitr_1.48          htmltools_0.5.8.1  </span></span>
<span><span class="co">## [10] rmarkdown_2.28      lifecycle_1.0.4     cli_3.6.3          </span></span>
<span><span class="co">## [13] sass_0.4.9          pkgdown_2.1.0       textshaping_0.4.0  </span></span>
<span><span class="co">## [16] jquerylib_0.1.4     systemfonts_1.1.0   compiler_4.4.1     </span></span>
<span><span class="co">## [19] tools_4.4.1         ragg_1.3.2          bslib_0.8.0        </span></span>
<span><span class="co">## [22] evaluate_0.24.0     yaml_2.3.10         BiocManager_1.30.25</span></span>
<span><span class="co">## [25] jsonlite_1.8.8      rlang_1.1.4         fs_1.6.4</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
