<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Doublet identifiation in single-cell ATAC-seq • scDblFinder</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Doublet identifiation in single-cell ATAC-seq">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDblFinder</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.19.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/scDblFinder.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/computeDoubletDensity.html">Scoring potential doublets from simulated densities</a></li>
    <li><a class="dropdown-item" href="../articles/findDoubletClusters.html">Detecting clusters of doublet cells with DE analyses</a></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Introduction to the scDblFinder package</a></li>
    <li><a class="dropdown-item" href="../articles/recoverDoublets.html">Recovering intra-sample doublets</a></li>
    <li><a class="dropdown-item" href="../articles/scATAC.html">Doublet identifiation in single-cell ATAC-seq</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/plger/scDblFinder/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Doublet identifiation in single-cell ATAC-seq</h1>
                        <h4 data-toc-skip class="author">Pierre-Luc
Germain</h4>
            <address class="author_afil">
      University and ETH Zürich<br><small class="dont-index">Source: <a href="https://github.com/plger/scDblFinder/blob/devel/vignettes/scATAC.Rmd" class="external-link"><code>vignettes/scATAC.Rmd</code></a></small>
      <div class="d-none name"><code>scATAC.Rmd</code></div>
    </address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>An introduction to the methods implemented for doublet
      detection in single-cell ATAC-seq.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Analyses in single-cell RNAseq are typically limited to a relatively
small (e.g. one or two thousands) set of features that are most
informative; these are often the genes with a higher expression (and
hence more chances of being quantified). In contrast, single-cell
ATACseq (scATACseq) data is considerably more sparse, with most reads
being spread across hundreds of thousands of regions. In this context,
selecting a subset of genes is highly ineffective, and therefore many of
the methods developed for single-cell RNAseq are not easily applicable,
and need to be adapted. Methods have therefore been developed
specifically for scATACseq data (Granja et al. 2021; Thibodeau et
al. 2021).</p>
<p>This vignette presents different approaches to doublet detection in
single-cell ATAC-seq implemented in the package: the first is an
adaptation of <code>scDblFinder</code>, the second a reimplementation of
the AMULET method from Thibodeau et al. (2021). The latter has the
advantage of capturing homotypic doublets, but does not perform well in
all datasets, and especially requires the cells to have a high library
size. We therefore next present two ways of combining the two.</p>
</div>
<div class="section level2">
<h2 id="applying-the-scdblfinder-method">Applying the scDblFinder method<a class="anchor" aria-label="anchor" href="#applying-the-scdblfinder-method"></a>
</h2>
<p>With default parameters, the <code>scDblFinder</code> method performs
very poorly on scATACseq data due to the increase spread of the reads
across many features. Since working with all features (i.e. tiles or
peaks) is computationally very expensive, an alternative approach is to
begin by reducing the size of the dataset, not through
<em>selection</em> (as in scRNAseq), but by <em>aggregating</em>
correlated features into a relatively small set. This has the advantage
of using all information, as well as making the count data more
continuous. This method yields comparable performance to specialized
single-cell ATACseq software (Germain et al., 2021).</p>
<p>The feature aggregation can be triggered using the
<code>aggregateFeatures=TRUE</code> argument, which will aggregate peak
or tile counts into the number of meta-features defined by the
<code>nfeatures</code>. If the number of meta-features is low (which we
recommend), the meta-features can be directly used to calculated
distances rather than going through the SVD step (which can be triggered
with the <code>processing</code> argument). Such an example would
be:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/plger/scDblFinder" class="external-link">scDblFinder</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># we use a dummy SingleCellExperiment as example:</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mockDoubletSCE.html">mockDoubletSCE</a></span><span class="op">(</span>ngenes<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span>
<span><span class="co"># setting low number of artificial doublets (same as ncells) just for speedup:</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scDblFinder.html">scDblFinder</a></span><span class="op">(</span><span class="va">sce</span>, artificialDoublets<span class="op">=</span><span class="fl">1</span>, aggregateFeatures<span class="op">=</span><span class="cn">TRUE</span>, nfeatures<span class="op">=</span><span class="fl">25</span>, processing<span class="op">=</span><span class="st">"normFeatures"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Aggregating features...</span></span></code></pre>
<pre><code><span><span class="co">## Creating ~526 artificial doublets...</span></span></code></pre>
<pre><code><span><span class="co">## Dimensional reduction</span></span></code></pre>
<pre><code><span><span class="co">## Evaluating kNN...</span></span></code></pre>
<pre><code><span><span class="co">## Training model...</span></span></code></pre>
<pre><code><span><span class="co">## iter=0, 17 cells excluded from training.</span></span></code></pre>
<pre><code><span><span class="co">## iter=1, 34 cells excluded from training.</span></span></code></pre>
<pre><code><span><span class="co">## iter=2, 36 cells excluded from training.</span></span></code></pre>
<pre><code><span><span class="co">## Threshold found:0.705</span></span></code></pre>
<pre><code><span><span class="co">## 18 (3.4%) doublets called</span></span></code></pre>
<p>If you encounter problems running the aggregation-based approach on
large datasets, first make sure you have the <code>mbkmeans</code>
package installed.</p>
</div>
<div class="section level2">
<h2 id="using-the-amulet-method">Using the Amulet method<a class="anchor" aria-label="anchor" href="#using-the-amulet-method"></a>
</h2>
<p>The AMULET method from Thibodeau et al. (2021) is based on the
assumption that, in a diploid cell, any given genomic region should be
captured at most twice. Therefore, cells with loci covered by more than
two fragments are indicative of the droplet being a doublet. Of note,
this approach has the advantage of capturing homotypic doublets, which
instead tend to be missed by other methods. Since it was only available
in the form of a mixture of java and python scripts, we re-implemented
the method in <code>scDblFinder</code> (see <code><a href="../reference/amulet.html">?amulet</a></code>),
leading to equal or superior results to the original implementation
(Germain et al. 2021).</p>
<p>As in the original implementation, we recommend excluding the
mitochondrial and sex chromosomes, as well as repetitive regions. This
can be specified with the <code>regionsToExclude</code> argument (see
the underlying <code><a href="../reference/getFragmentOverlaps.html">?getFragmentOverlaps</a></code>). It can be used as
follows:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># here we use a dummy fragment file for example:</span></span>
<span><span class="va">fragfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_fragments.tsv.gz"</span>, package<span class="op">=</span><span class="st">"scDblFinder"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># we might also give a GRanges of repeat elements, so that these regions are excluded:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">repeats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr6"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">1000</span>,<span class="fl">2000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># it's better to combine these with mitochondrial and sex chromosomes</span></span>
<span><span class="va">otherChroms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"M"</span>,<span class="st">"chrM"</span>,<span class="st">"MT"</span>,<span class="st">"X"</span>,<span class="st">"Y"</span>,<span class="st">"chrX"</span>,<span class="st">"chrY"</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">1L</span>,width<span class="op">=</span><span class="fl">10</span><span class="op">^</span><span class="fl">8</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># here since I don't know what chromosome notation you'll be using I've just put them all,</span></span>
<span><span class="co"># although this will trigger a warning when combining them:</span></span>
<span><span class="va">toExclude</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">repeats</span>, <span class="va">otherChroms</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># we then launch the method</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/amulet.html">amulet</a></span><span class="op">(</span><span class="va">fragfile</span>, regionsToExclude<span class="op">=</span><span class="va">toExclude</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Fragment file is not tabix-indexed, requiring thewhole file to be imported in memory.</span></span></code></pre>
<pre><code><span><span class="co">## 11:00:01 - Splitting and subsetting barcodes...</span></span></code></pre>
<pre><code><span><span class="co">## 11:00:01 - Obtaining overlaps...</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">##          nFrags uniqFrags nAbove2 total.nAbove2     p.value     q.value</span></span>
<span><span class="co">## barcode1    878       878       1             1 0.475069053 0.791781755</span></span>
<span><span class="co">## barcode2   2401      2401       0             0 0.798103482 0.798103482</span></span>
<span><span class="co">## barcode3   2325      2325       1             1 0.475069053 0.791781755</span></span>
<span><span class="co">## barcode4   1882      1882       0             0 0.798103482 0.798103482</span></span>
<span><span class="co">## barcode5   1355      1355       6             6 0.001335761 0.006678806</span></span></code></pre>
<p>The results is a data.frame with statistics for each barcode,
including a p-value. In contrast to the <code>scDblFinder</code> score,
a lower p-value here is indicative of the droplet being more likely to
be a doublet (as in the original method). By default, only the barcodes
with a minimum number of reads are considered, but it is possible to
specify the droplets for which to gather statistics using the
<code>barcodes</code> argument.</p>
<p>While the package includes an implementation that works based on
peak/tile count matrices (see <code><a href="../reference/amuletFromCounts.html">?amuletFromCounts</a></code>), it has a
much lower performance with respect to the one based directly on the
fragment files (see <code><a href="../reference/amulet.html">?amulet</a></code>), and we therefore discourage
its use.</p>
<p>The workhorse behind the <code>amulet</code> function is the
<code>getFragmentOverlaps</code>, which also includes all of the
relevant arguments. If the fragment files are not Tabix-indexed, the
whole fragment file will have to be loaded in memory for processing;
while this ensures relatively rapid computation, it has high memory
requirements. Therefore, if the fragment file is Tabix-indexed (as is
for instance done as part of the ArchR pipeline), it will be read and
processed per chromosome, which is a little slower due to overhead, but
keeps memory requirements rather low. This behavior can be disabled by
specifying <code>fullInMemory=TRUE</code>.</p>
</div>
<div class="section level2">
<h2 id="combining-mehtods">Combining mehtods<a class="anchor" aria-label="anchor" href="#combining-mehtods"></a>
</h2>
<p>While the <code>scDblFinder</code>-based approach generally performs
well, none of the two approach is optimal across all datasets tested. We
therefore investigated two strategies for combining the rationales of
each approach.</p>
<p>The Amulet method tends to perform best with datasets that have
homotypic doublets and where cells have a high library size (i.e. median
library size per cell of 10-15k reads), while the
<code>scDblFinder</code>-based approach works better for heterotypic
doublets. Until an optimal solution is found, we recommend using
multiple approaches to inform decisions, in particular using the p-value
combination method below.</p>
<div class="section level3">
<h3 id="the-clamulet-method">The Clamulet method<a class="anchor" aria-label="anchor" href="#the-clamulet-method"></a>
</h3>
<p>The <code>clamulet</code> method (Classification-powered Amulet-like
method) operates similarly to the <code>scDblFinder</code> method, but
generates artificial doublets by operating on the fragment coverages.
This has the advantage that the number of loci covered by more than two
reads can be computed for artificial doublets, enabling the use of this
feature (along with the kNN-based ones) in a classification scheme. It
however has the disadvantage of being rather slow and memory hungry, and
appears to be outperformed by a simple p-value combination of the two
methods (see below). We therefore <em>do not</em> recommend its
usage.</p>
<p>The <code>clamulet</code> method uses the aforementioned aggregation
approach, and its usage includes a number of arguments from both the
<code>scDblFinder</code> and <code>amulet</code> method (see in
particular <code><a href="../reference/getFragmentOverlaps.html">?getFragmentOverlaps</a></code>):</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># not run</span></span>
<span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clamulet.html">clamulet</a></span><span class="op">(</span><span class="st">"path/to/fragments.tsv.gz"</span><span class="op">)</span></span></code></pre></div>
<p>Since our dummy fragment file is so small (5 barcodes), here we’ll
have to adjust the arguments for an example to run:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clamulet.html">clamulet</a></span><span class="op">(</span><span class="va">fragfile</span>, k<span class="op">=</span><span class="fl">2</span>, nfeatures<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 11:00:02 - Reading full fragments...</span></span></code></pre>
<pre><code><span><span class="co">## 11:00:02 - Splitting and subsetting barcodes...</span></span></code></pre>
<pre><code><span><span class="co">## 11:00:02 - Computing coverages</span></span></code></pre>
<pre><code><span><span class="co">## 11:00:02 - Obtaining windows</span></span></code></pre>
<pre><code><span><span class="co">## 11:00:03 - Obtaining window counts</span></span></code></pre>
<pre><code><span><span class="co">## 11:00:03 - Aggregating features</span></span></code></pre>
<pre><code><span><span class="co">## Warning in (function (A, nv = 5, nu = nv, maxit = 1000, work = nv + 7, reorth =</span></span>
<span><span class="co">## TRUE, : You're computing too large a percentage of total singular values, use a</span></span>
<span><span class="co">## standard svd instead.</span></span></code></pre>
<pre><code><span><span class="co">## 11:00:03 - Computing features for artificial doublets</span></span></code></pre>
<pre><code><span><span class="co">## 11:00:03 - Counting overlaps for real cells</span></span></code></pre>
<pre><code><span><span class="co">## 11:00:03 - Counting overlaps for artificial doublets</span></span></code></pre>
<pre><code><span><span class="co">## 11:00:03 - Scoring network</span></span></code></pre>
<pre><code><span><span class="co">## 11:00:03 - Iterative training</span></span></code></pre>
<pre><code><span><span class="co">## iter=0, 0 cells excluded from training.</span></span></code></pre>
<pre><code><span><span class="co">## iter=1, 0 cells excluded from training.</span></span></code></pre>
<pre><code><span><span class="co">## 11:00:03 Done!</span></span></code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span></span></code></pre></div>
<pre><code><span><span class="co">##          total nAbove2 total.nAbove2  weighted ratio.k2 include.in.training</span></span>
<span><span class="co">## barcode1    20       1             1 0.5843011      0.5                TRUE</span></span>
<span><span class="co">## barcode2    14       0             0       NaN      0.5                TRUE</span></span>
<span><span class="co">## barcode3    11       1             1 1.0000000      1.0                TRUE</span></span>
<span><span class="co">## barcode4    12       0             0       NaN      0.5                TRUE</span></span>
<span><span class="co">## barcode5    18       6             6 0.4368783      0.5                TRUE</span></span>
<span><span class="co">##              score</span></span>
<span><span class="co">## barcode1 0.4174298</span></span>
<span><span class="co">## barcode2 0.4174298</span></span>
<span><span class="co">## barcode3 0.4174298</span></span>
<span><span class="co">## barcode4 0.4174298</span></span>
<span><span class="co">## barcode5 0.4174298</span></span></code></pre>
<p>The score can then be interpreted as for <code>scDblFinder</code>. We
however note that this method proved <em>inferior to
alternatives</em>.</p>
</div>
<div class="section level3">
<h3 id="simple-p-value-combination">Simple p-value combination<a class="anchor" aria-label="anchor" href="#simple-p-value-combination"></a>
</h3>
<p>The amulet and scDblFinder scores above can be simply combined by
treating them as p-values and aggregating them (here using Fisher’s
method from the <em><a href="https://CRAN.R-project.org/package=aggregation" class="external-link">aggregation</a></em>
package, but see also the <em><a href="https://CRAN.R-project.org/package=metap" class="external-link">metap</a></em>
package):</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span><span class="op">$</span><span class="va">scDblFinder.p</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>, <span class="st">"scDblFinder.score"</span><span class="op">]</span></span>
<span><span class="va">res</span><span class="op">$</span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"scDblFinder.p"</span>, <span class="st">"p.value"</span><span class="op">)</span><span class="op">]</span>, <span class="fl">1</span>, FUN<span class="op">=</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">x</span><span class="op">[</span><span class="va">x</span><span class="op">&lt;</span><span class="fl">0.001</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0.001</span> <span class="co"># prevent too much skew from very small or 0 p-values</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu">aggregation</span><span class="fu">::</span><span class="fu">fisher</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>We found this to perform better than averaging the scores or their
ranks, and while it is not the very best method in any of the datasets
tested, it has a more robust performance overall (see Germain et al.,
2021).</p>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Jeffrey M. Granja et al., “ArchR Is a Scalable Software Package for
Integrative Single-Cell Chromatin Accessibility Analysis,” Nature
Genetics, February 25, 2021, 1–9, <a href="https://doi.org/10.1038/s41588-021-00790-6" class="external-link uri">https://doi.org/10.1038/s41588-021-00790-6</a></p>
<p>Asa Thibodeau et al., “AMULET: A Novel Read Count-Based Method for
Effective Multiplet Detection from Single Nucleus ATAC-Seq Data,” Genome
Biology 22, no. 1 (December 2021): 252, <a href="https://doi.org/10.1186/s13059-021-02469-x" class="external-link uri">https://doi.org/10.1186/s13059-021-02469-x</a></p>
<p>Pierre-Luc Germain et al., “Doublet Identification in Single-Cell
Sequencing Data Using ScDblFinder” (F1000Research, September 28, 2021),
<a href="https://doi.org/10.12688/f1000research.73600.1" class="external-link uri">https://doi.org/10.12688/f1000research.73600.1</a></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] scDblFinder_1.19.2          SingleCellExperiment_1.26.0</span></span>
<span><span class="co">##  [3] SummarizedExperiment_1.34.0 Biobase_2.64.0             </span></span>
<span><span class="co">##  [5] GenomicRanges_1.56.1        GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">##  [7] IRanges_2.38.1              S4Vectors_0.42.1           </span></span>
<span><span class="co">##  [9] BiocGenerics_0.50.0         MatrixGenerics_1.16.0      </span></span>
<span><span class="co">## [11] matrixStats_1.4.0           BiocStyle_2.32.1           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] bitops_1.0-8              gridExtra_2.3            </span></span>
<span><span class="co">##   [3] rlang_1.1.4               magrittr_2.0.3           </span></span>
<span><span class="co">##   [5] scater_1.32.1             compiler_4.4.1           </span></span>
<span><span class="co">##   [7] DelayedMatrixStats_1.26.0 systemfonts_1.1.0        </span></span>
<span><span class="co">##   [9] vctrs_0.6.5               pkgconfig_2.0.3          </span></span>
<span><span class="co">##  [11] crayon_1.5.3              fastmap_1.2.0            </span></span>
<span><span class="co">##  [13] XVector_0.44.0            scuttle_1.14.0           </span></span>
<span><span class="co">##  [15] utf8_1.2.4                Rsamtools_2.20.0         </span></span>
<span><span class="co">##  [17] rmarkdown_2.28            UCSC.utils_1.0.0         </span></span>
<span><span class="co">##  [19] ggbeeswarm_0.7.2          ragg_1.3.2               </span></span>
<span><span class="co">##  [21] xfun_0.47                 bluster_1.14.0           </span></span>
<span><span class="co">##  [23] zlibbioc_1.50.0           cachem_1.1.0             </span></span>
<span><span class="co">##  [25] beachmat_2.20.0           jsonlite_1.8.8           </span></span>
<span><span class="co">##  [27] DelayedArray_0.30.1       BiocParallel_1.38.0      </span></span>
<span><span class="co">##  [29] irlba_2.3.5.1             parallel_4.4.1           </span></span>
<span><span class="co">##  [31] cluster_2.1.6             R6_2.5.1                 </span></span>
<span><span class="co">##  [33] bslib_0.8.0               limma_3.60.4             </span></span>
<span><span class="co">##  [35] rtracklayer_1.64.0        xgboost_1.7.8.1          </span></span>
<span><span class="co">##  [37] jquerylib_0.1.4           Rcpp_1.0.13              </span></span>
<span><span class="co">##  [39] bookdown_0.40             knitr_1.48               </span></span>
<span><span class="co">##  [41] Matrix_1.7-0              igraph_2.0.3             </span></span>
<span><span class="co">##  [43] tidyselect_1.2.1          abind_1.4-5              </span></span>
<span><span class="co">##  [45] yaml_2.3.10               viridis_0.6.5            </span></span>
<span><span class="co">##  [47] codetools_0.2-20          curl_5.2.2               </span></span>
<span><span class="co">##  [49] lattice_0.22-6            tibble_3.2.1             </span></span>
<span><span class="co">##  [51] evaluate_0.24.0           desc_1.4.3               </span></span>
<span><span class="co">##  [53] Biostrings_2.72.1         pillar_1.9.0             </span></span>
<span><span class="co">##  [55] BiocManager_1.30.25       generics_0.1.3           </span></span>
<span><span class="co">##  [57] RCurl_1.98-1.16           ggplot2_3.5.1            </span></span>
<span><span class="co">##  [59] sparseMatrixStats_1.16.0  munsell_0.5.1            </span></span>
<span><span class="co">##  [61] scales_1.3.0              glue_1.7.0               </span></span>
<span><span class="co">##  [63] metapod_1.12.0            tools_4.4.1              </span></span>
<span><span class="co">##  [65] BiocIO_1.14.0             data.table_1.16.0        </span></span>
<span><span class="co">##  [67] BiocNeighbors_1.22.0      ScaledMatrix_1.12.0      </span></span>
<span><span class="co">##  [69] locfit_1.5-9.10           GenomicAlignments_1.40.0 </span></span>
<span><span class="co">##  [71] fs_1.6.4                  scran_1.32.0             </span></span>
<span><span class="co">##  [73] XML_3.99-0.17             grid_4.4.1               </span></span>
<span><span class="co">##  [75] edgeR_4.2.1               colorspace_2.1-1         </span></span>
<span><span class="co">##  [77] GenomeInfoDbData_1.2.12   beeswarm_0.4.0           </span></span>
<span><span class="co">##  [79] BiocSingular_1.20.0       restfulr_0.0.15          </span></span>
<span><span class="co">##  [81] vipor_0.4.7               cli_3.6.3                </span></span>
<span><span class="co">##  [83] rsvd_1.0.5                textshaping_0.4.0        </span></span>
<span><span class="co">##  [85] fansi_1.0.6               S4Arrays_1.4.1           </span></span>
<span><span class="co">##  [87] viridisLite_0.4.2         dplyr_1.1.4              </span></span>
<span><span class="co">##  [89] gtable_0.3.5              sass_0.4.9               </span></span>
<span><span class="co">##  [91] digest_0.6.37             SparseArray_1.4.8        </span></span>
<span><span class="co">##  [93] ggrepel_0.9.5             dqrng_0.4.1              </span></span>
<span><span class="co">##  [95] rjson_0.2.22              htmltools_0.5.8.1        </span></span>
<span><span class="co">##  [97] pkgdown_2.1.0             lifecycle_1.0.4          </span></span>
<span><span class="co">##  [99] httr_1.4.7                statmod_1.5.0            </span></span>
<span><span class="co">## [101] MASS_7.3-60.2</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
